ads/google/a4a/traffic-experiments.js

```diff
@@ -22,12 +22,101 @@
+ *
+ * If we're selected into the overall traffic experiment, this function will
+ * also attach an experiment or control branch ID to the `Element` as
+ * a side-effect.
+ *
+ * @param {!Window} win  Host window for the ad.
+ * @param {!Element} element Ad tag Element.
+ * @param {string} experimentId ID for the experiment to use.
+ * @param {!Branches} experiment and control branch IDs to use.
+ * @return {boolean}  Whether Google Ads should attempt to render via the A4A
+ *   pathway.
+ */
+export function googleAdsIsA4AEnabled(win, element, experimentId, branches) {
+  // Note: Theoretically, isProxyOrigin is the right way to do this, b/c it
+  // will be kept up to date with known proxies.  However, it doesn't seem to
+  // be compatible with loading the example files from localhost.  To hack
+  // around that, just say that we're A4A eligible if we're in local dev
+  // mode, regardless of origin path.
+  if (isGoogleAdsA4AValidEnvironment(win)) {
+    // Page is served from a supported domain.
+    handleUrlParameters(win, experimentId, branches);
+    const experimentInfo = {};
+    experimentInfo[experimentId] = branches;
+    setupPageExperiments(win, experimentInfo);
+    if (isExperimentOn(win, experimentId)) {
+      // Page is selected into the overall traffic experiment.
+      if (getPageExperimentBranch(win, experimentId) === branches.experiment) {
+        // Page is on the "experiment" (i.e., use A4A rendering pathway)
+        // branch of the overall traffic experiment.
+        addExperimentIdToElement(branches.experiment, element);
+        return true;
+      } else {
+        // Page is on the "control" (i.e., use traditional, 3p iframe
+        // rendering pathway) branch of the overall traffic experiment.
+        addExperimentIdToElement(branches.control, element);
+        return false;
+      }
+    }
+  }
+  // Serving location doesn't qualify for A4A treatment or page is not in the
+  // traffic experiment.
+  return false;
+}
+
+/**
+ * Set experiment state from URL parameter, if present.
+ *
+ * @param {!Window} win
+ * @param {string} experimentId
+ * @param {!Branches} branches
+ * @visibleForTesting
+ */
+function handleUrlParameters(win, experimentId, branches) {
+  const a4aParam = /(?:\?|&)a4a=([0-9]+)/.exec(win.location.search);
+  if (a4aParam && a4aParam.length > 1) {
+    switch (a4aParam[1]) {
+      case '0':
+        // Not selected into experiment.  Disable the experiment altogether, so
+        // that setupPageExperiments doesn't accidentally enable it.
+        forceExperimentBranch(win, experimentId, false);
+        break;
+      case '1':
+        // Selected in; on control branch.
+        forceExperimentBranch(win, experimentId, branches.control);
+        break;
+      case '2':
+        // Selected in; on experiment branch.
+        forceExperimentBranch(win, experimentId, branches.experiment);
+        break;
+      default:
+        dev.warn('a4a-config', 'Unknown a4a URL parameter: ',
```

Comment:
I don't think so.  The ad should still function in that case, via the 3p iframe path.  It just won't attach an experiment ID.  But that's okay -- most traffic won't be part of the experiment regardless.